river/expert/bandit.py

```diff
@@ -0,0 +1,269 @@
+        # average reward instead of cumulated (otherwise favors arms which are pulled often)
+        return np.argmax(self._average_reward)
+
+    @property
+    def best_model(self):
+        return self[self._best_model_idx]
+
+    @property
+    def percentage_pulled(self):
+        percentages = self._N / sum(self._N)
+        return percentages
+
+    def predict_one(self, x):
+        best_arm = self._pull_arm()
+        y_pred = self._pred_func(self[best_arm])(x)
+        return y_pred
+
+    def learn_one(self, x, y):
+        chosen_arm = self._pull_arm()
+        chosen_model = self[chosen_arm]
+
+        y_pred = chosen_model.predict_one(x)
+        self.metric.update(y_pred=y_pred, y_true=y)
+        chosen_model.learn_one(x=x, y=y)
+
+        # Update bandit internals (common to all bandit)
+        reward = self._compute_scaled_reward(y_pred=y_pred, y_true=y)
+        self._n_iter += 1
+        self._N[chosen_arm] += 1
+        self._average_reward[chosen_arm] += (1.0 / self._N[chosen_arm]) * \
+                                            (reward - self._average_reward[chosen_arm])
+
+        # Specific update of the arm for certain bandit class
+        self._update_arm(chosen_arm, reward)
+
+        if self.print_every:
+            if (self._n_iter % self.print_every) == 0:
+                self._print_info()
+
+        if self.save_percentage_pulled:
+            self.store_percentage_pulled += [self.percentage_pulled]
+
+        if self.save_metric_values:
+            self.metric_values += [self.metric._eval(y_pred, y)]
+
+        return self
+
+    def add_models(self, new_models):
+        if not isinstance(new_models, list):
+            raise TypeError("Argument `new_models` must be of a list")
+
+        length_new_models = len(new_models)
+        # Careful, not validation of the model is done here (contrary to __init__)
+        self.models += new_models
+        self._n_arms += length_new_models
+        self._N = np.concatenate([self._N, np.zeros(length_new_models, dtype=np.int)])
+        self._average_reward = np.concatenate(
+            [self._average_reward, np.zeros(length_new_models, dtype=np.float)]
+        )
+
+    def _compute_scaled_reward(self, y_pred, y_true, update_scaler=True):
+        metric_value = self.metric._eval(y_pred, y_true)
+        metric_to_reward_dict = {
+            "metric": metric_value if self.metric.bigger_is_better else (-1) * metric_value
+        }
+        if update_scaler:
+            self.reward_scaler.learn_one(metric_to_reward_dict)
+        reward = self.reward_scaler.transform_one(metric_to_reward_dict)["metric"]
+        return reward
+
+    def _print_info(self):
+        print(
+            str(self),
+            str(self.metric),
+            "Best model id: " + str(self._best_model_idx),
+            sep="\n\t"
+        )
+
+
+class EpsilonGreedyBandit(Bandit):
+    def __init__(self, epsilon=0.1, epsilon_decay=None, **kwargs):
+        super().__init__(**kwargs)
```

Comment:
This part seems to make the tests fail (same part for UCBBandit). Should I enumerate all the parameters (that are in `Bandit.__init__`) instead of using `**kwargs` ?

Reply:
Yes, you should :)

Reply:
Actually I try with it but the tests do fail again. I see also that expert.SuccessiveHalving* are purposely exclude from the test (in `get_all_estimators`) while `expert.EWARegressor` is not. Is the kind of regressor I implemented supposed to pass this test at all ?

Reply:
You're right, it's not supposed to pass the tests by default. You need implement a `_default_params` method, like what is done [here](https://github.com/online-ml/river/blob/d3e56400a0ece7113f454d26d46fe18139a19a21/river/multiclass/ovo.py#L73). This method will get called [here](https://github.com/online-ml/river/blob/master/river/test_.py#L96) to initialize the state of your model.

Reply:
Also, I just add default LinearRegression in a Pipeline and tests pass. It somehow surprise me since I didn't gave default arg for `metric` and `reward_scaler`.

**Not related to this PR** but one test failed : 

```python
======================================================================== FAILURES ========================================================================
___________________________________________________ [doctest] river.neighbors.sam_knn.SAMKNNClassifier ___________________________________________________
052     >>> from river import metrics
053     >>> from river import neighbors
054 
055     >>> dataset = synth.ConceptDriftStream(position=500, width=20, seed=1).take(1000)
056 
057     >>> model = neighbors.SAMKNNClassifier(window_size=100)
058 
059     >>> metric = metrics.Accuracy()
060 
061     >>> evaluate.progressive_val_score(dataset, model, metric)
Expected:
    Accuracy: 59.90%
Got:
    Accuracy: 58.20%

/path/to/package/river/neighbors/sam_knn.py:61: DocTestFailure
```

Reply:
@etiennekintzler any reason why you put the linear regression in a pipeline? Yes it's a bit weird that the tests are passing if those are required arguments :/. Could be worth investigating.

@smastelini any idea why the above test is failing?

Reply:
@MaxHalford Because most of the time I've used `models` that were `Pipeline` (that includes scaler and regressor), but actually it should work the same if I just put regressor.

Reply:
I do confirm that it works without putting it in Pipeline (as it should since I just use the `predict_one` and `learn_one` which are defined for both Pipeline and Regressors).

Reply:
> @smastelini any idea why the above test is failing?

I could not reproduce that failure in my local setup (in the `master` branch). Pinging @jacobmontiel, since he has more experience with this specific k-NN variant than I do.

Reply:
@MaxHalford @smastelini 

> I just add default LinearRegression in a Pipeline and tests pass. It somehow surprise me since I didn't gave default arg for metric and reward_scaler.

It passes on my local setup using `pytest` but it did fail on the CI pipeline (https://travis-ci.org/github/online-ml/river/jobs/745694837). It fails because I gave the wrong name for the argument name. If it had gone further it would have failed as well since I didn't provide default argument for parameters `metric` and `reward_scaler`

Reply:
> > @smastelini any idea why the above test is failing?
> 
> I could not reproduce that failure in my local setup (in the `master` branch). Pinging @jacobmontiel, since he has more experience with this specific k-NN variant than I do.

FYU the `master` branch in my fork is 10 commits behind `master` in `online-ml/river` however these 10 commits are mostly doc so..

Reply:
> It passes on my local setup using pytest but it did fail on the CI pipeline (https://travis-ci.org/github/online-ml/river/jobs/745694837).

Very weird! Not sure what to say. Are you sure your development setup is correct?

Reply:
@MaxHalford 

My bad :/ I forget that I add the bandits classes in the `ignored` tuple in `get_all_estimators`.

I can now confirm that it fails as in the CI when I do **not** provide the arguments.

Reply:
> > @smastelini any idea why the above test is failing?

> I could not reproduce that failure in my local setup (in the master branch). Pinging @jacobmontiel, since he has more experience with this specific k-NN variant than I do.

I am confused, is this failing?

Reply:
I clone and rebuild the whole package (following https://github.com/online-ml/river/blob/master/CONTRIBUTING.md#installation) and it fails at the same place:

```python
======================================================================== FAILURES ========================================================================
___________________________________________________ [doctest] river.neighbors.sam_knn.SAMKNNClassifier ___________________________________________________
052     >>> from river import metrics
053     >>> from river import neighbors
054 
055     >>> dataset = synth.ConceptDriftStream(position=500, width=20, seed=1).take(1000)
056 
057     >>> model = neighbors.SAMKNNClassifier(window_size=100)
058 
059     >>> metric = metrics.Accuracy()
060 
061     >>> evaluate.progressive_val_score(dataset, model, metric)
Expected:
    Accuracy: 59.90%
Got:
    Accuracy: 58.20%

/home/etienne/PROJETS/river_test/river/river/neighbors/sam_knn.py:61: DocTestFailure

```

 I am using Python 3.7.7 and Ubuntu 20.04.1 LTS. 

I can open an issue to keep it out of the PR